File: bms/go/go-gt/QAP_SolveSA/QAP_SolveSA.go
package test

import (
	"fmt"
	"math"
	"math/rand"
)

var (
	Verbose bool
)

const ()

func QAP_SolveSA(a *Matrix, b *Matrix, p Vector, m int64) int64 {
	var i int64
	n := p.Len()
	w := make(Vector, n)
	w.Copy(p)
	cc := cost(a, b, p)
	c, dmin, dmax := initQAP(a, b, w, cc)
	var t0 float64 = float64(dmin + (dmax-dmin)/10.0)
	tf := float64(dmin)
	beta := (t0 - tf) / (float64(m) * t0 * tf)
	var fail int64 = 0
	tries := n * (n - 1) / 2
	tfound := t0
	var temp float64 = t0
	var r int64 = 0
	var s int64 = 1

	// SA iterations
	for i = 0; i < m; i++ {
		temp /= (beta*temp + 1)
		s++
		if s >= n {
			r++
			if r >= n-1 {
				r = 0
			}
			s = r + 1
		}
		d := delta(a, b, w, r, s)
		if (d < 0) || (rand.Float64() < math.Exp(-float64(d)/temp)) || (fail == tries) {
			c += d
			w.Swap(r, s)
			fail = 0
		} else {
			fail++
		}
		if fail == tries {
			beta = 0
			temp = tfound
		}

		// Best solution improved ?
		if c < cc {
			cc = c
			p.Copy(w)
			tfound = temp
			if Verbose {
				fmt.Printf("iteration %d: cost=%d\n", i, cc)
				p.Print()
			}
		}
	}
	return cc
}

func (v Vector) Len() int64 {
	return int64(len(v))
}

func (v Vector) Copy(w Vector) {
	for i := 0; i < len(v); i++ {
		v[i] = w[i]
	}
}

func cost(a *Matrix, b *Matrix, p Vector) (c int64) {
	var i, j int64
	c = 0
	for i = 0; i < p.Len(); i++ {
		for j = 0; j < p.Len(); j++ {
			c += a.Get(i, j) * b.Get(p[i], p[j])
		}
	}
	return c
}

func (m Matrix) Get(i int64, j int64) int64 {
	return m.A[i*m.N+j]
}

func initQAP(a *Matrix, b *Matrix, w Vector, c int64) (int64, int64, int64) {
	var (
		dmin, dmax int64
	)
	n := w.Len()
	for i := 0; i < 10000; i++ {
		r := rand.Int63n(n)
		s := rand.Int63n(n - 1)
		if s >= r {
			s = s + 1
		}
		d := delta(a, b, w, r, s)
		c += d
		dmin = min(dmin, d)
		dmax = max(dmax, d)
		w.Swap(r, s)
	}
	return c, dmin, dmax
}

func delta(a *Matrix, b *Matrix, p Vector, r int64, s int64) (d int64) {
	var i int64
	d = int64((a.Get(r, r)-a.Get(s, s))*(b.Get(p[s], p[s])-b.Get(p[r], p[r])) +
		(a.Get(r, s)-a.Get(s, r))*(b.Get(p[s], p[r])-b.Get(p[r], p[s])))
	for i = 0; i < p.Len(); i++ {
		if i != r && i != s {
			d += (a.Get(i, r)-a.Get(i, s))*(b.Get(p[i], p[s])-b.Get(p[i], p[r])) +
				(a.Get(r, i)-a.Get(s, i))*(b.Get(p[s], p[i])-b.Get(p[r], p[i]))
		}
	}
	return d
}

func min(a, b int64) int64 {
	if a < b {
		return a
	}
	return b
}

func max(a, b int64) int64 {
	if a > b {
		return a
	}
	return b
}

func (p Vector) Swap(i int64, j int64) {
	x := p[i]
	p[i] = p[j]
	p[j] = x
}

func (v Vector) Print() {
	for i := 0; i < len(v); i++ {
		fmt.Printf("%d ", v[i])
	}
	fmt.Print("\n")
}

type Vector []int64

type Matrix struct {
	N int64
	A []int64
}

File: translations_only/success/go/go-gt/QAP_SolveSA/claude2/Restart/QAP_SolveSA.rs

use std::fmt;

struct Matrix {
    n: i64,
    a: Vec<i64>,
}

type Vector = Vec<i64>;

fn qap_solve_sa(a: &Matrix, b: &Matrix, p: Vector, m: i64) -> i64 {
    let mut i = 0;
    let n = p.len() as i64;
    let mut w = p.clone();
    let mut cc = cost(&a, &b, &w);
    let (mut c, dmin, dmax) = init_qap(&a, &b, &mut w, &mut cc);
    let mut t0 = (dmin + (dmax - dmin) / 10) as f64;
    let tf = dmin as f64;
    let mut beta = (t0 - tf) / (m as f64 * t0 * tf);
    let mut fail = 0;
    let tries = n * (n - 1) / 2;
    let mut tfound = t0;
    let mut temp = t0;
    let mut r = 0;
    let mut s = 1;

    while i < m {
        temp /= beta * temp + 1.0;
        s += 1;
        if s >= n {
            r += 1;
            if r >= n - 1 {
                r = 0;
            }
            s = r + 1;
        }
        let d = delta(&a, &b, &w, r, s);
        if (d < 0) || (rand::random::<f64>() < (-d as f64).exp() / temp) || (fail == tries) {
            c += d;
            swap(&mut w, r, s);
            fail = 0;
        } else {
            fail += 1;
        }
        if fail == tries {
            beta = 0.0;
            temp = tfound;
        }

        if c < cc {
            cc = c;
            copy(&mut w, &p);
            tfound = temp;
            if verbose() {
                println!("iteration {}: cost={}", i, cc);
                print_vector(&p);
            }
        }

        i += 1;
    }

    cc
}

fn len(v: &Vector) -> i64 {
    v.len() as i64
}

fn copy(v: &mut Vector, w: &Vector) {
    v.clone_from(w);
}

fn cost(a: &Matrix, b: &Matrix, p: &Vector) -> i64 {
    let mut c = 0;
    for i in 0..p.len() as i64 {
        for j in 0..p.len() as i64 {
            c += a.get(i, j) * b.get(p[i as usize], p[j as usize]);
        }
    }
    c
}

impl Matrix {
    fn get(&self, i: i64, j: i64) -> i64 {
        self.a[i as usize * self.n as usize + j as usize]
    }
}

fn init_qap(a: &Matrix, b: &Matrix, w: &mut Vector, c: &mut i64) -> (i64, i64, i64) {
    let mut dmin = 0;
    let mut dmax = 0;
    let n = w.len() as i64;
    for _ in 0..10000 {
        let r = rand::random::<i64>() % n;
        let s = rand::random::<i64>() % (n - 1);
        let d = delta(&a, &b, &w, r, s);
        *c += d;
        dmin = min(dmin, d);
        dmax = max(dmax, d);
        swap(w, r, s);
    }
    (*c, dmin, dmax)
}

fn delta(a: &Matrix, b: &Matrix, p: &Vector, r: i64, s: i64) -> i64 {
    let mut d = (a.get(r, r) - a.get(s, s)) * (b.get(p[s as usize], p[s as usize]) - b.get(p[r as usize], p[r as usize]))
        + (a.get(r, s) - a.get(s, r)) * (b.get(p[s as usize], p[r as usize]) - b.get(p[r as usize], p[s as usize]));
    for i in 0..p.len() as i64 {
        if i != r && i != s {
            d += (a.get(i, r) - a.get(i, s)) * (b.get(p[i as usize], p[s as usize]) - b.get(p[i as usize], p[r as usize]))
                + (a.get(r, i) - a.get(s, i)) * (b.get(p[s as usize], p[i as usize]) - b.get(p[r as usize], p[i as usize]));
        }
    }
    d
}

fn min(a: i64, b: i64) -> i64 {
    if a < b { a } else { b }
}

fn max(a: i64, b: i64) -> i64 {
    if a > b { a } else { b }
}

fn swap(v: &mut Vector, i: i64, j: i64) {
    v.swap(i as usize, j as usize);
}

fn print_vector(v: &Vector) {
    for i in v {
        print!("{} ", i);
    }
    println!("");
}

static mut VERBOSE: bool = false;

fn verbose() -> bool {
    unsafe { VERBOSE }
}

